### ISSUE TITLE
[Payment card / Subscription] make `TransferPolicyOwnership` 1:1:1 and move as much code as possible to Auth

### ISSUE THREAD

**Issue by blimpich** [Expensify/App#46994]
# Problem
TransferPolicyOwnership is not 1:1:1. We strive for all our commands to be 1:1:1 and in addition to the inherent benefit of making this command 1:1:1 this will also unblock https://github.com/Expensify/App/issues/46933 which is necessary in order to create SCA supported flow to add a billing card while transferring ownership.

# Solution

Make TransferPolicyOwnership 1:1:1 and move as much of the web code as possible to auth.

**Comment by blimpich** [comment]
Working on this today.

**Comment by blimpich** [comment]
Note: even though the command in web is `TransferPolicyOwnership` in auth its `TakePolicyOwnership`. That tripped me up.

Update: made a [draft PR](https://github.com/Expensify/Auth/pull/11946), slow work, lots of details. Trying my best to divide it up into smaller and smaller bite sized pieces.

**Comment by blimpich** [comment]
Difficulty moving some of the calls to auth, confused about forwarding commands and using them in Auth. Though seems like [this is what I want](https://expensify.slack.com/archives/C03TQ48KC/p1720516783946489?thread_ts=1720472105.705149&cid=C03TQ48KC). Synchronous and able to check for errors.

**Comment by blimpich** [comment]
Made good progress today. I got my first "draft" of the changes done for the auth PR, same for the web PR (which is much simpler), and I also was able to save myself some work by figuring out that there was code in web that was no longer needed  and could be deleted entirely instead of being ported into Auth ([PR](https://github.com/Expensify/Web-Expensify/pull/43072)).

Still lots of work to be done.

**Comment by blimpich** [comment]
Not overdue. Was out on Friday and then it was the weekend. Working on this today.

**Comment by blimpich** [comment]
Got it compiling without errors! Now to test it out manually and then also probably writing some tests / assessing what automated tests already cover this functionality.

**Comment by blimpich** [comment]
Fixed some of the auth tests but not all. Some seems to be quite odd and hard to figure out why they are failing. Pushed up my updates branches with the work I have done today.

**Comment by blimpich** [comment]
Spent a little more time and didn't fix any tests, but I did get it to work manually after making one more small change to my web-expensify PR. Logging off for now, will continue work on this tomorrow.

**Comment by blimpich** [comment]
Fixed all the auth tests, now need to fix the broken web tests.

**Comment by blimpich** [comment]
Chugging along. Fixed some of the web tests but not all of them. Very difficult to debug and find all the errors, but I did find quite a few of them. Some incorrect onyx updates, bad messages on Auth responses, and just some straight up typos/bugs that I missed on my first pass. Biggest revelation was that I need to know whether the request is coming from new dot or not in order to be able to get the web tests to pass. I thought I could get around that but it seems excruciatingly difficult to do so. So I made [this PR](https://github.com/Expensify/Web-Expensify/pull/43139) to help me with that.

Progress is being made. I'm hoping this PR will be ready for review by the end of the week but its possible it won't be.

**Comment by blimpich** [comment]
Down to one last broken test in the web tests. So close!
![Screenshot 2024-08-15 at 2 07 51 PM](https://github.com/user-attachments/assets/49d07f1f-f5e4-486d-aafc-be323e24b32b)

**Comment by blimpich** [comment]
Alright I was hoping to avoid this but we need to temporarily change some functionality. The `TransferOwnershipTest::testTransferOwnershipNewDotOwnerAmountOwed` test has proved to me that there is a flow that we simply can't make work fully until we've deleted the php code that will be replaced by the auth port.

The issue is that the php code checks "does the user taking over have an amount owed, if so error out, if they don't and the owner has an amount owed, then transfer it over." So we then want the auth code to just go through and transfer the policy over like normal, but now that the amount owed has been transferred Auth now sees that the new owner does have an amount owed, and therefore is errors out when actually we want it to go through like nothing was wrong!

We're stuck in a position where we either let people who want to take over policies who already have amount owed's temporarily transfer policies or we block people from transferring policies where the owner has an amount owed. This will only be an issue temporarily, and I think I can make it such that we get all the necessary changes in one deploy.

I think the way to do this the most smoothly will be to temporarily allow people to transfer policies even though they have an amount owed.

**Comment by blimpich** [comment]
Okay here is what is left to do on this:
- [ ] get green build on [Auth PR](https://github.com/Expensify/Auth/pull/11946) (I think I have this, went green locally)
- [ ] get green build locally using [Web-E feature branch](https://github.com/Expensify/Web-Expensify/pull/43074)
- [ ] cleanup [Auth PR](https://github.com/Expensify/Auth/pull/11946)
- [ ] put [Auth PR](https://github.com/Expensify/Auth/pull/11946) out for review
- [ ] get [Auth PR](https://github.com/Expensify/Auth/pull/11946) merged
- [ ] immediately get [Web-PR](https://github.com/Expensify/Web-Expensify/pull/43074) merged
- [ ] create PR and do this issue to finalize changes: https://github.com/Expensify/App/issues/47528

**Comment by blimpich** [comment]
Half day today, didn't make progress. I'm stumped as to why this 1 web-integration test is passing locally and failing in GH. Will come back Monday and work on it.

**Comment by blimpich** [comment]
Working on this today.

**Comment by blimpich** [comment]
Okay I think I finally figured out why that last test isn't passing on the Auth PR. The integration tests we run for Auth PRs are based off the production web branch, and the production branch is waiting for changes to be deployed that we need in order to get a green build. See comment [here](https://github.com/Expensify/Auth/pull/11946#issuecomment-2299489009).

So waiting for those changes to be deployed, meanwhile I'll work on cleaning up the PR and getting the web-PR ready for review.

**Comment by blimpich** [comment]
Green build on [Auth PR](https://github.com/Expensify/Auth/pull/11946) 🎉, working on getting green build locally using the [web feature branch](https://github.com/Expensify/Web-Expensify/pull/43074), 1 failing test left.

**Comment by blimpich** [comment]
Got a green build locally on the web feature branch! 🎉

Lot of clean up to be done but [the checklist](https://github.com/Expensify/App/issues/46994#issuecomment-2292444538) is getting completed.

**Comment by blimpich** [comment]
Got green build locally for the feature branch, 2nd checkbox is done

**Comment by blimpich** [comment]
Lots of progress today, got the auth pr out for review and cleaned up both the auth PR and web PR. Waiting for feedback now.

**Comment by blimpich** [comment]
Good progress today, got my initial round of feedback on the PR from two reviewers, working through that feedback now.

**Comment by blimpich** [comment]
Not overdue, just responded to feedback, working on getting green build again and getting approval.

**Comment by blimpich** [comment]
Not overdue, almost done with this. See checklist. Will put a PR up for the last checklist item today. Will take awhile to get a green build though I think because we will probably have to wait for our web changes to get deployed to production before we'll be able to get a green build on the web-integration tests.

**Comment by blimpich** [comment]
Almost there, web changes just got deployed to production, attempting to get a green build on the last little auth PR [here](https://github.com/Expensify/Auth/pull/12274)

**Comment by blimpich** [comment]
Got green build, PR is out for review

**Comment by blimpich** [comment]
Closing since there is a whole other issue dedicated to tracking the [final checkbox](https://github.com/Expensify/App/issues/46994#issuecomment-2292444538) of this issue and all other work is done

**Comment by blimpich** [comment]
renamed because we didn't make the command fully 1:1:1, we just made it more 1:1:1, enough to unblock other issues
